### ISSUE TITLE
Why does   `numerator(const cpp_rational&)`  not return a `const& cpp_int`

### ISSUE THREAD

**Issue by afabri** [boostorg/multiprecision#377]
I would expect not get a copy. Is the rational number more than just two integers?

**Comment by afabri** [comment]
Would it make sense to have a traits class which tells us that it is const&   or a copy?

**Comment by jzmaddock** [comment]
Not really, as of current develop, none of the rational types store pairs of `number<>` to which we could return a reference.  Some might be able to be type-punned to such a type, but it would be deeply risky IMO.  Others like mpq_rational definitely don't have anything we can return a reference to.

**Comment by jzmaddock** [comment]
I'm leaning towards agreeing with you on this - but it will require a fair bit of refactoring - so please bare with me!

**Comment by afabri** [comment]
Did this question get any further thought?  Could that be an interesting Google Summer of Code topic co-mentored by the CGAL project and you @jzmaddock ?

**Comment by jzmaddock** [comment]
Sorry haven't had a chance yet, a joint SOC project isn't a bad idea, but for this change the student would need to spend a lot of time understanding our internals, for what is actually a relative small amount of coding.

Question, do you guys need to be able to write:

```
numerator(x) = 2;
```

If not, then expression templates would work and avoid the type-punning.  However, given that the type punning issue only effects `mpq_rational` we could just hold our noses and pun away ;)  I somewhat worry about painting ourselves into a corner when dealing with whatever future types may come along though.

**Comment by jzmaddock** [comment]
The referenced PR addresses this.  Note that there are some rational types where numerator/denominator still return by value, this is to support for example `debug_adaptor` where the returned type is unrelated to the stored type and adds additional state.  @afabri is this good enough to fix things at your end?
